gce external LB: add a function to verify the requested IP address

yujuhong opened the pull request:
Factor out the logic for verifying the user-requested IP for better
readability and testing. Also rename a few variables for clarity.

yujuhong committed:
gce external LB: add a function to verify the requested IP address

yujuhong commented:
This should have minimal (if any) functional change. I'd like to get this in before adding more code.

nicksardo commented on pkg/cloudprovider/providers/gce/gce_addresses_fakes.go:
```diff
@@ -19,12 +19,14 @@ package gce
 )
 
 type FakeCloudAddressService struct {
+	sync.RWMutex
```
What's the point of the lock if each unit test is instantiating its own instance?

yujuhong replied:
This is not needed for the PR, but I think in general having a lock is still useful as we add more tests.If you rather wait until the lock is needed, I can also remove it.

nicksardo commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external_test.go:
```diff
@@ -37,3 +42,44 @@ func TestEnsureStaticIP(t *testing.T) {
+		if tc.ipList != nil {
+			s.SetRegionalAddresses(region, tc.ipList)
+		}
+		isUserOwnedIP, err := verifyUserRequestedIP(s, region, tc.requestedIP, tc.fwdRuleIP, lbRef)
```
How about [subtests](https://blog.golang.org/subtests)?

yujuhong replied:
Sure. Updated.

nicksardo commented on pkg/cloudprovider/providers/gce/gce_addresses_fakes.go:
```diff
@@ -34,14 +36,33 @@ type FakeCloudAddressService struct {
+		cas.reservedAddrs[ip] = true
+		regionStore[name] = &compute.Address{Name: name, Address: ip}
+	}
+}
```
Recommend checking that the IP/address-name doesn't exist before creating the map entries and returning a potential error. Mistakes in unit tests do happen and would be nice to catch them then.

yujuhong replied:
I changed the unit test to reset the map before adding new addresses.

nicksardo commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -456,6 +422,39 @@ func (gce *GCECloud) DeleteExternalTargetPoolAndChecks(name, region, clusterID s
 }
 
+// verifyUserRequestedIP checks the user-provided IP  to see whether it can be
+// used for the LB.  It also returns whetherthe IP is considered owned by the
```
sp whetherthe

yujuhong replied:
Done

nicksardo commented:
I didn't spot any logic errors, but should get a few more eyeballs on it.

yujuhong commented:
Thanks for the review! Addressed the comments.

> I didn't spot any logic errors, but should get a few more eyeballs on it.

Feel free to assign more reviewers as you see fit.

FengyunPan commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -456,6 +422,39 @@ func (gce *GCECloud) DeleteExternalTargetPoolAndChecks(name, region, clusterID s
+	if requestedIP == fwdRuleIP {
+		// The requested IP is not a static IP, but is currently assigned
+		// to this forwarding rule, so we can just use it.
+		return false, nil
```
Is it 'return true, nil'?

yujuhong replied:
Nope. See the original code [here](https://github.com/kubernetes/kubernetes/blob/58ea4e41d4947c0eff0b59685b29d86b3cbe50ae/pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go#L138-L147)

yujuhong commented:
/cc @thockin @freehan @MrHohn to see if anyone else wants to review

yujuhong commented:
Ping @nicksardo @bowei for another round of review

MrHohn commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -110,75 +110,41 @@ func (gce *GCECloud) ensureExternalLoadBalancer(clusterName, clusterID string, a
+	if !isUserOwnedIP {
+		// If we are not using the user-owned IP, either promote the
+		// emphemeral IP used by the fwd rule, or create a new static IP.
+		ipAddr, existed, err := ensureStaticIP(gce, loadBalancerName, serviceName.String(), gce.region, fwdRuleIP)
```
Log the ensured IP for debuggability?

yujuhong replied:
Done.

yujuhong commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -456,6 +422,39 @@ func (gce *GCECloud) DeleteExternalTargetPoolAndChecks(name, region, clusterID s
+		glog.V(4).Infof("verifyUserRequestedIP: the requested static IP %q (name: %s) for LB %s exists.", requestedIP, existingAddress.Name, lbRef)
+		return true, nil
+	}
+	if requestedIP == fwdRuleIP {
```
Done.

yujuhong reviewed (commented):
@MrHohn, added more logging as you suggested.

yujuhong commented:
Addressed all comments.

yujuhong commented:
/retest

yujuhong commented:
/retest

bowei commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -456,6 +423,40 @@ func (gce *GCECloud) DeleteExternalTargetPoolAndChecks(name, region, clusterID s
+	// case we shouldn't delete it anyway).
+	existingAddress, err := s.GetRegionAddressByIP(region, requestedIP)
+	if err != nil && !isNotFound(err) {
+		return false, fmt.Errorf("failed to check if the static IP %q exists: %v", requestedIP, err)
```
I'm not sure wrapping the error makes sense here. It seems to make more sense to glog.Errorf and pass the original error up.

yujuhong replied:
Sure. Done.

bowei commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -456,6 +423,40 @@ func (gce *GCECloud) DeleteExternalTargetPoolAndChecks(name, region, clusterID s
+	// ephemeral IP (or it was a different static IP owned by the user, in which
+	// case we shouldn't delete it anyway).
+	existingAddress, err := s.GetRegionAddressByIP(region, requestedIP)
+	if err != nil && !isNotFound(err) {
```
seems to make the logic clearer if we call out isNotFound case explicitly

```
if isNotFound(err) {
  ...
}
if err != nil {
  ...
}
```

yujuhong replied:
If `isNotFound(err)` is true, the code would continue to check whether `requestedIP == fwdRuleIP`.  Changing this may not make the logic more readable.

bowei commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -110,75 +110,42 @@ func (gce *GCECloud) ensureExternalLoadBalancer(clusterName, clusterID string, a
 		if isSafeToReleaseIP {
 			if err := gce.DeleteRegionAddress(loadBalancerName, gce.region); err != nil && !isNotFound(err) {
-				glog.Errorf("failed to release static IP %s for load balancer (%v(%v), %v): %v", ipAddress, loadBalancerName, serviceName, gce.region, err)
+				glog.Errorf("failed to release static IP %s for load balancer (%v(%v), %v): %v", ipAddressToUse, loadBalancerName, serviceName, gce.region, err)
```
[minor] Capitalize Failed

bowei commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -110,75 +110,42 @@ func (gce *GCECloud) ensureExternalLoadBalancer(clusterName, clusterID string, a
 		}
 		if isSafeToReleaseIP {
 			if err := gce.DeleteRegionAddress(loadBalancerName, gce.region); err != nil && !isNotFound(err) {
```
seems cleaner to have:
```
err := gce.DeleteRegionAddress(loadBalancerName, gce.region)
if isNotFound(err) {
}
if err != nil {
}
...
```

bowei commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -110,75 +110,42 @@ func (gce *GCECloud) ensureExternalLoadBalancer(clusterName, clusterID string, a
 			}
 		} else {
-			glog.Warningf("orphaning static IP %s during update of load balancer (%v(%v), %v): %v", ipAddress, loadBalancerName, serviceName, gce.region, err)
+			glog.Warningf("orphaning static IP %s during update of load balancer (%v(%v), %v): %v", ipAddressToUse, loadBalancerName, serviceName, gce.region, err)
```
promote this to an Error.

bowei commented on pkg/cloudprovider/providers/gce/gce_loadbalancer_external.go:
```diff
@@ -456,6 +423,40 @@ func (gce *GCECloud) DeleteExternalTargetPoolAndChecks(name, region, clusterID s
 	return nil
 }
 
+// verifyUserRequestedIP checks the user-provided IP  to see whether it can be
```
isUserOwnedIP seems like a better name?

yujuhong replied:
Do you mean to renamed `ownedByUser` to `isUserOwnedIP`? Done. Although it's not used anywhere, and it's just here for clarity.

yujuhong commented:
@bowei addressed the comments.

nicksardo commented:
/lgtm

nicksardo commented:
/approve no-issue

yujuhong commented:
Squashed the commits. Will re-apply the labels.
